### DISCUSSION TITLE
How does proot run non-startable programs?

### DISCUSSION THREAD

**Discussion by Maxython** [termux/termux-packages#8677] [General]
Hi all.
I recently decided to do an experiment with glibc (the purpose of the experiment is to run it without `proot` or with minimal use of it).  Yes, I know that android does not support glibc and mixing glibc with bionic is a bad idea, but hey, no one forbids compiling it under termux prefix separately from bionic.

In short, I compiled this miracle, but there was one problem that I did not expect.  When launching any glibc program, it says `Bad system call`.  I ran the program under `strace` and the error was shown to be due to `syscall set_robust_list`.

Using libc.so.6 as an example:
```bash
$ strace ./libc.so.6
execve("./libc.so.6", ["./libc.so.6"], 0x7fce2d81e0 /* 23 vars */) = 0
brk(NULL)                               = 0x5c1a99a000
faccessat(AT_FDCWD, "/etc/ld.so.preload", R_OK) = -1 ENOENT (No such file or directory)
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x70f0ab3000
set_tid_address(0x70f0ab35a0)           = 29753
set_robust_list(0x70f0ab35b0, 24)       = 485074089392
--- SIGSYS {si_signo=SIGSYS, si_code=SYS_SECCOMP, si_call_addr=0x70f0a97f24, si_syscall=__NR_set_robust_list, si_arch=AUDIT_ARCH_AARCH64} ---
+++ killed by SIGSYS +++
Bad system call
```

After a little googling, I realized that the linux android kernel does not have this function, then I decided to run glibc programs under `proot` for fun, and lo and behold, they worked. 

After that, I had a question, what does `proot` do to make non-starting commands run and is it possible to implement this in the command itself?

**Comment by ghost** [comment]
the `bad system call` occurred because of seccomp triggered and yes probably missing libc functions. usually non-trusted binaries which was not compiled in Android's libc

`proot` on the other hand intercepts system calls which allows execution of any binaries and preventing this to happen. this is a common workaround for `bad system call` errors

**Reply by Maxython** [reply]
Is it for all syscalls or for specific?

**Reply by ghost** [reply]
all possible syscalls registered by `proot` otherwise a program would still run without `bad system call` error but it would spit `function not implemented` or the program's actual error

**Comment by Maxython** [comment]
The result of the experiment: glibc works on termux without proot, the most important thing is to configure it.
